File: server/sockets/history.ts
import type { FinalServer, FinalSocket, SocketCallback } from "$/types";
import type { ScanEntry, ScanStatus } from "@/types";
import { atomicWrite } from "$/lib/utils";
import { publicDir } from "$/persist";
import { getPermissions } from "@/lib/permission";

const historyFilePath = publicDir("output", "history.json");

async function readHistoryFromFile() {
    try {
        const historyFile = Bun.file(historyFilePath);
        if (await historyFile.exists()) {
            const json = await historyFile.json();
            return json as ScanEntry[];
        }
    } catch (error) {
        console.error("Error reading history file:", error);
    }
    return [];
}

async function writeHistoryToFile(history: ScanEntry[]) {
    try {
        await atomicWrite(historyFilePath, JSON.stringify(history, null, 2));
    } catch (error) {
        console.error("Error writing to history file:", error);
    }
}

let scanHistory: ScanEntry[] = await readHistoryFromFile();
let isWriting = false;
let needsWriteAgain = false;

async function syncHistoryToDisk() {
    if (isWriting) {
        needsWriteAgain = true;
        return;
    }

    isWriting = true;

    try {
        await writeHistoryToFile(scanHistory);
    } catch (err) {
        console.error("Disk write failed: ", err);
    }

    isWriting = false;

    // If mutations happened while writing, run it one more time with the latest data
    if (needsWriteAgain) {
        needsWriteAgain = false;
        await syncHistoryToDisk();
    }
}

export function history(io: FinalServer, socket: FinalSocket) {
    socket.on("client:history:init", () => {
        if (!socket.data.user) return;
        socket.emit("server:history:update", scanHistory);
    });

    socket.on(
        "client:history:validation",
        (qrData: string, status: ScanStatus, callback: SocketCallback<string>) => {
            const user = socket.data.user;
            if (!user || !getPermissions(user.authorizeLevel).canScan) {
                return callback({
                    status: "error",
                    error: `Unauthorized validation attempt by user: ${user?.name}`,
                });
            }

            let duplicatedValidator = "";
            const isDuplicate = scanHistory.some((entry) => {
                if (entry.data === qrData && entry.status === "Valid") {
                    duplicatedValidator = entry.validatorName;
                    return true;
                }

                return false;
            });
            if (isDuplicate) {
                return callback({
                    status: "info",
                    message: `This entry data (${qrData}) has already been validated by ${duplicatedValidator}`,
                });
            }

            const newScan: ScanEntry = {
                id: `scan_${Date.now()}`,
                data: qrData,
                status: status,
                validatorName: user.name,
                validatedAt: new Date().toISOString(),
            };
            scanHistory.unshift(newScan);
            void syncHistoryToDisk();
            callback({
                status: "success",
                data: `Validation for ${qrData} has been submitted`,
            });
            io.emit("server:history:update", scanHistory);
        },
    );

    socket.on("client:history:delete", (idToDelete: string) => {
        const user = socket.data.user;
        if (!user || !getPermissions(user.authorizeLevel).canDelete) {
            console.log(`Unauthorized delete attempt by user:`, user?.name);
            return;
        }

        const initialLength = scanHistory.length;
        scanHistory = scanHistory.filter((entry) => entry.id !== idToDelete);
        if (scanHistory.length < initialLength) {
            void syncHistoryToDisk();
            io.emit("server:history:update", scanHistory);
        }
    });
}

File: server/sockets/history.go
package sockets

import (
	"fmt"
	"premark/db"
	"premark/lib"
	"premark/types"

	"github.com/zishang520/socket.io/servers/socket/v3"
)

// Mutex to protect read-then-write (check-and-insert) race conditions across concurrent sockets.
var validationMu = lib.NewKeyedRWMutex()

func checkIfDuplicate(projectId string, creatorHash []byte, rowId int, allowDuplicateValid bool, maxValidDuplicate int) (string, int) {
	var threshold int
	if allowDuplicateValid {
		threshold = maxValidDuplicate
	} else {
		threshold = 1
	}

	batchNumber := getCachedBatchNumber(lib.BytesToBase64(creatorHash))
	if batchNumber == 0 {
		return "", 0
	}

	exists, err := db.HasValidPresenceAtLeast(projectId, batchNumber, rowId, threshold)
	if err != nil || !exists {
		return "", 0
	}

	if allowDuplicateValid {
		msg := fmt.Sprintf("This entry row (%d) has already been validated more than %d times", rowId, maxValidDuplicate)
		return msg, 1
	}

	msg := fmt.Sprintf("This entry row (%d) has already been validated", rowId)
	return msg, 2
}

func fetchAndEmitHistory(io *socket.Server, client *socket.Socket, creatorHash []byte, projectId string, broadcast bool) {
	dataset, err := db.FindDatasetByProjectId(creatorHash, projectId, false)
	if err != nil {
		return
	}

	batchNumber := getCachedBatchNumber(lib.BytesToBase64(creatorHash))
	if batchNumber == 0 {
		return
	}

	histories, err := db.FindPresenceHistoriesByProjectId(projectId, batchNumber, dataset.Key, true)
	if err != nil {
		return
	}

	if broadcast {
		io.Emit("server:history:update", histories)
	} else if client != nil {
		client.Emit("server:history:update", histories)
	}
}

func registerHistoryHandlers(io *socket.Server, client *socket.Socket) {
	client.On("client:history:init", func(args ...any) {
		ctx := client.Data().(*types.SocketData)
		if ctx.User == nil {
			return
		}

		pID, creatorHash, err := getActiveProject(ctx.UserHashBytes, ctx.UserHashBase64)
		if err != nil {
			return
		}

		fetchAndEmitHistory(io, client, creatorHash, pID, false)
	})

	client.On("client:history:check:duplicate", func(args ...any) {
		if len(args) < 4 {
			return
		}
		qrData, _ := args[0].(string)
		allowDuplicateValid, _ := args[1].(bool)
		maxValidDuplicate, err := tryParseInt(args[2])
		if err != nil {
			invokeAck(args, types.SocketResponse{Status: "error", Error: "Invalid maximum valid duplicate."})
			return
		}

		ctx := client.Data().(*types.SocketData)
		if ctx.User == nil || !GetPermissions(ctx.User.AuthorizeLevel).CanScan {
			msg := fmt.Sprintf("Unauthorized check attempt by user: %s", ctx.User.Name)
			invokeAck(args, types.SocketResponse{Status: "error", Error: msg})
			return
		}

		decrypted, err := decryptUserData(qrData)
		if err != nil {
			invokeAck(args, types.SocketResponse{Status: "error", Error: "Decryption failed."})
			return
		}

		rowHash, err := lib.CreateSearchHash(decrypted)
		if err != nil {
			invokeAck(args, types.SocketResponse{Status: "error", Error: "Hashing failed."})
			return
		}

		pID, creatorBytes, err := getActiveProject(ctx.UserHashBytes, ctx.UserHashBase64)
		if err != nil {
			msg := fmt.Sprintf("Unable to get active project: %s", err.Error())
			invokeAck(args, types.SocketResponse{Status: "error", Error: msg})
			return
		}

		rowId, err := db.FindDatasetRowId(pID, rowHash)
		if err != nil || rowId == -1 {
			msg := fmt.Sprintf("Row not found: %s", rowHash)
			invokeAck(args, types.SocketResponse{Status: "error", Error: msg})
		}

		validationMu.DoRLock(fmt.Sprintf("%s:%d", pID, rowId), func() {
			_, duplicatedCode := checkIfDuplicate(pID, creatorBytes, rowId, allowDuplicateValid, maxValidDuplicate)

			res := map[string]any{
				"decrypted":      decrypted,
				"duplicatedCode": duplicatedCode,
			}
			invokeAck(args, types.SocketResponse{Status: "success", Data: res})
		})
	})

	client.On("client:history:validation", func(args ...any) {
		if len(args) < 3 {
			return
		}
		qrData, _ := args[0].(string)
		status, _ := args[1].(string)

		ctx := client.Data().(*types.SocketData)
		if ctx.User == nil || !GetPermissions(ctx.User.AuthorizeLevel).CanScan {
			msg := fmt.Sprintf("Unauthorized validation attempt by user: %s", ctx.User.Name)
			invokeAck(args, types.SocketResponse{Status: "error", Error: msg})
			return
		}

		rowHash, err := lib.CreateSearchHash(qrData)
		if err != nil {
			invokeAck(args, types.SocketResponse{Status: "error", Error: "Hashing failed."})
			return
		}

		var (
			allowDuplicateValid bool
			maxValidDuplicate   int
		)
		pID, creatorBytes, err := getActiveProject(ctx.UserHashBytes, ctx.UserHashBase64)
		if err != nil {
			msg := fmt.Sprintf("Unable to get active project: %s", err.Error())
			invokeAck(args, types.SocketResponse{Status: "error", Error: msg})
			return
		}

		project, _ := db.FindProjectScanOptById(creatorBytes, pID)
		if project != nil {
			if v, ok := project["allowDuplicateValid"].(bool); ok {
				allowDuplicateValid = v
			}
			if v, ok := project["maxValidDuplicate"].(int); ok {
				maxValidDuplicate = v
			}
		}

		rowId, err := db.FindDatasetRowId(pID, rowHash)
		if err != nil || rowId == -1 {
			msg := fmt.Sprintf("Row not found: %s", rowHash)
			invokeAck(args, types.SocketResponse{Status: "error", Error: msg})
			return
		}

		batchNumber := getCachedBatchNumber(lib.BytesToBase64(creatorBytes))
		if batchNumber == 0 {
			invokeAck(args, types.SocketResponse{Status: "error", Error: fmt.Sprintf("Failed to get batch number for project: %s", pID)})
			return
		}

		// Mutex lock prevents TOCTOU race conditions when multiple workers scan concurrently
		hasDuplicates, err := func() (bool, error) {
			unlock := validationMu.Lock(fmt.Sprintf("%s:%d", pID, rowId))
			defer unlock()

			duplicatedMsg, _ := checkIfDuplicate(pID, creatorBytes, rowId, allowDuplicateValid, maxValidDuplicate)
			if duplicatedMsg != "" {
				invokeAck(args, types.SocketResponse{Status: "info", Message: duplicatedMsg})
				return true, nil
			}

			_, err = db.AddPresenceHistory(pID, rowId, ctx.UserHashBytes, status, batchNumber)
			return false, err
		}()

		if hasDuplicates {
			return
		}
		if err != nil {
			invokeAck(args, types.SocketResponse{Status: "error", Error: "Failed to record scan."})
			return
		}

		msg := fmt.Sprintf("Validation for %s has been submitted", qrData)
		invokeAck(args, types.SocketResponse{Status: "success", Data: msg})

		// Broadcast updated history to all connected sockets
		fetchAndEmitHistory(io, nil, creatorBytes, pID, true)
	})

	client.On("client:history:delete", func(args ...any) {
		if len(args) < 1 {
			return
		}
		idToDelete, _ := args[0].(string)
		ctx := client.Data().(*types.SocketData)
		if ctx.User == nil || !GetPermissions(ctx.User.AuthorizeLevel).CanDelete {
			return
		}

		pID, creatorHash, err := getActiveProject(ctx.UserHashBytes, ctx.UserHashBase64)
		if err != nil {
			return
		}

		if err := db.DeletePresenceHistory(idToDelete); err != nil {
			return
		}

		fetchAndEmitHistory(io, nil, creatorHash, pID, true)
	})
}
